src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+    1 value is missing. If False, remove rows where all values are missing.
+
+    Returns:
+    - pd.DataFrame. The preprocessed dataset
+    """
+    try:
+        initial_length = len(data)
+        if remove_if_all_na:  # remove if everything is NA
+            data = data.dropna(how="all")
+        else:
+            data = data.dropna()
+        dropped_length = len(data)
+
+        # Filter orders with >= 5 items
+        num_items_ordered = data.groupby("order_id")["outcome"].transform("sum")
+        processed_data = data[num_items_ordered >= num_items]
+
+        logging.info(f"Length initial data: {len(data)}")
+        logging.info(f"Rows dropped with NA's: {initial_length - dropped_length}")
+        logging.info(f"Length filtered data: {len(processed_data)}\n")
+
+        return processed_data
+
+    except FileNotFoundError as e:
+        logging.error(f"Error: File not found. Details {e}")
+        return
+
+
+def split_sets(df: pd.DataFrame, label: str) -> (pd.DataFrame, pd.Series):
+    """
+    Return a df with X set (features) and a series y set (outcome)
+
+    Parameters:
+    - df: pd.DataFrame
+
+    Returns:
+    - X: pd.DataFrame (features only)
+    - y: pd.Series (label to predict)
+    """
+    X = df.drop(columns=label)
+    y = df[label]
+
+    return X, y
+
+
+def temporal_data_split(
+    data: pd.DataFrame,
+    validation_days: int = 10,
+    test_days: int = 10,
+    label: str = "outcome",
+    cols: List[str] = None,
+):
+    """
+    Perform the temporal data splitting into train, validation, and test set.
+
+    Parameters:
+    - data : pd.DataFrame
+    - validation_days (int): Number of days for the validation set (default: 10).
+    - test_days (int): Number of days for the test set (default: 10).
+    - label (str): Name of the outcome variable (default: 'outcome').
+
+    Returns:
+    - pd.DataFrame: Training set features.
+    - pd.Series: Training set outcome.
+    - pd.DataFrame: Validation set features.
+    - pd.Series: Validation set outcome.
+    - pd.DataFrame: Test set features.
+    - pd.Series: Test set outcome.
+    """
+    try:
+        # Check order_date is a datetime and then we group the orders
+        try:
+            data["order_date"] = pd.to_datetime(data["order_date"]).dt.date
+        except KeyError as ke:
+            logging.error(f"Key Error: {str(ke)}")
+            return
+        daily_orders = data.groupby("order_date").order_id.nunique()
+
+        start_date = daily_orders.index.min()
+        end_date = daily_orders.index.max()
+        logging.info(f"Date from: {start_date}")
+        logging.info(f"Date to: {end_date}")
+
+        # Based on the number of days, we get the train days
+        num_days = len(daily_orders)
+        train_days = num_days - test_days - validation_days
+
+        # Train
+        train_start = daily_orders.index.min()
+        train_end = train_start + timedelta(days=train_days)
```

Comment:
Nit pick, you are using `train_days+1` instead of `train_days`: if `train_start` is the 5 and `train_days=2`, then `train_end = 5+2 = 7`